### ISSUE TITLE
Wunused-const-variable warns when variable template is used in body of uninstantiated function template

### ISSUE THREAD

**Issue by davidstone** [llvm/llvm-project#49877]
|  |  |
| --- | --- |
| Bugzilla Link | [50533](https://llvm.org/bz50533) |
| Version | trunk |
| OS | Linux |
| CC | @dwblaikie,@DougGregor,@zygoloid |

## Extended Description 
The following translation unit

```
namespace {

template<typename T>
constexpr bool false_ = false;

}

template<typename T>
void f() {
	static_assert(false_<T>);
}
```

warns when compiled with `-Wunused-const-variable`:

```
<source>:4:16: warning: unused variable 'false_' [-Wunused-const-variable]
constexpr bool false_ = false;
               ^
1 warning generated.
Compiler returned: 0
```

even though the variable is clearly used.

See it live: https://godbolt.org/z/oYxMq1K6r

If an instantiation of `f` is visible, the warning goes away, but not all instantiations are necessarily visible. The warning is also triggered with the following slightly larger translation unit:

```
namespace {

template<typename T>
constexpr bool false_ = false;

}

template<typename T>
void f() {
	static_assert(false_<T>);
}

template<typename T>
concept c = requires { f<T>(); };

static_assert(c<int>);
```

See it live: https://godbolt.org/z/f74G5Tx54

**Comment by wheatman** [comment]
This was fixed between 12.0.1 and 13
https://godbolt.org/z/Yqrrj6oec

code
```
// compile with -Wunused-const-variable -std=c++20
namespace {

template<typename T>
constexpr bool false_ = false;

}

template<typename T>
void f() {
	static_assert(false_<T>);
}

template<typename T>
concept c = requires { f<T>(); };

static_assert(c<int>);
```

old warning
```
<source>:4:16: warning: unused variable 'false_' [-Wunused-const-variable]
constexpr bool false_ = false;
               ^
1 warning generated.
Compiler returned: 0
```
